shared/src/balancer/swap/fixed_point.rs

```diff
@@ -1,3 +1,307 @@
+//! Module emulating the operations on fixed points with exactly 18 decimals as
+//! used in the Balancer smart contracts. Their original implementation can be
+//! found at:
+//! https://github.com/balancer-labs/balancer-v2-monorepo/blob/6c9e24e22d0c46cca6dd15861d3d33da61a60b98/pkg/solidity-utils/contracts/math/FixedPoint.sol
+
 #![allow(dead_code)]
 
+use super::error::Error;
+use anyhow::{anyhow, bail};
+use ethcontract::U256;
+use lazy_static::lazy_static;
+use std::{
+    fmt::{self, Debug, Formatter},
+    str::FromStr,
+};
+
 mod logexpmath;
+
+#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+/// Fixed point numbers that represent exactly any rational number that can be
+/// represented with up to 18 decimals as long as it can be stored in 256 bits.
+/// It corresponds to Solidity's `ufixed256x18`.
+/// Operations on this type are implemented as in Balancer's FixedPoint library,
+/// including error codes, from which the name (Balancer Fixed Point).
+pub struct Bfp(U256);
+
+lazy_static! {
+    static ref ONE_18: U256 = U256::exp10(18);
+    static ref ZERO: Bfp = Bfp(U256::zero());
+    static ref EPSILON: Bfp = Bfp(U256::one());
+    static ref ONE: Bfp = Bfp(*ONE_18);
+    static ref MAX_POW_RELATIVE_ERROR: Bfp = Bfp(10000_usize.into());
+}
+
+impl From<usize> for Bfp {
+    fn from(num: usize) -> Self {
+        Self(U256::from(num).checked_mul(*ONE_18).unwrap())
+    }
+}
+
+impl FromStr for Bfp {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut split_dot = s.split('.');
+        let units = split_dot.next().unwrap();
+        let decimals = split_dot.next().unwrap_or("0");
+        if split_dot.next().is_some() {
+            bail!("Too many decimal separators");
+        }
+        if units.is_empty() || decimals.is_empty() || decimals.len() > 18 {
+            bail!("Invalid decimal representation");
+        }
+        Ok(Bfp(U256::from_dec_str(&format!("{:0<18}", decimals))?
```

Comment:
Does `{:0<18}` do anything? I think it left pads the value with zeroes so that it is always 18 characters long but this will not change the represented number.

Reply:
This pads with zeros to the right, it's needed because for example the number `1.01` is parsed as `decimals = 01`, and after padding to the left we get the right number of wei for the decimal: `format!("{:0<18}", decimals) = 010000000000000000`, which becomes `10000000000000000`. I'll add a comment about it.

Reply:
If it pads to the right then that makes perfect sense. I misremembered how the format string works.